rules/DowngradePhp70/Rector/FuncCall/DowngradeUncallableValueCallToCallUserFuncRector.php

```diff
@@ -0,0 +1,120 @@
+ */
+final class DowngradeUncallableValueCallToCallUserFuncRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Downgrade calling a value that is not directly callable in PHP 5 (property, static property, closure, …) to call_user_func.',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+final class Foo
+{
+    /** @var callable */
+    public $handler;
+    /** @var callable */
+    public static $staticHandler;
+}
+
+$foo = new Foo;
+($foo->handler)(/* args */);
+($foo::$staticHandler)(41);
+
+(function() { /* … */ })();
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class Foo
+{
+    /** @var callable */
+    public $handler;
+    /** @var callable */
+    public static $staticHandler;
+}
+
+$foo = new Foo;
+call_user_func($foo->handler, /* args */);
+call_user_func($foo::$staticHandler, 41);
+
+call_user_func(function() { /* … */ });
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [FuncCall::class];
+    }
+
+    /**
+     * @param FuncCall $node
+     */
+    public function refactor(Node $node): ?FuncCall
+    {
+        if ($node->name instanceof Name) {
+            return null;
+        }
+
+        if (! $this->isNotDirectlyCallableInPhp5($node->name) && ! $this->isLexedDifferentlyInPhp5($node->name)) {
+            return null;
+        }
+
+        $args = array_merge([new Arg($node->name)], $node->args);
+
+        return new FuncCall(new Name('call_user_func'), $args);
+    }
+
+    private function isNotDirectlyCallableInPhp5(Expr $expr): bool
+    {
+        // PHP 7 supports calling arbitrary callable expressions as `($expr)($args)`
+        // but PHP 5 does not like it.
+        return (
+            // Interpreted as MethodCall without parentheses.
+            $expr instanceof PropertyFetch
+            // Interpreted as StaticCall without parentheses.
+            || $expr instanceof StaticPropertyFetch
+            || $expr instanceof Closure
+            // The first function call does not even need to be wrapped in parentheses
+            // but PHP 5 still does not like curried functions like `f($args)($moreArgs)`.
+            || $expr instanceof FuncCall
+        );
+    }
+
+    private function isLexedDifferentlyInPhp5(Expr $expr): bool
+    {
+        return (
+            // PHP 5.6 parses `self::$foo['test']()` as a static method call
+            // `self::{$foo['test']}()` rather than `(self::$foo)['test']()` like PHP 7.0 does.
+            // See php5-static-property-weirdness.php.inc fixture.
+            $expr instanceof ArrayDimFetch && $expr->var instanceof StaticPropertyFetch
```

Comment:
Same parsing difference occurs with `self::$foo["test"]["foo"]();`, which could be captured by `$expr instanceof ArrayDimFetch && $expr->var instanceof ArrayDimFetch && $expr->var->var instanceof StaticPropertyFetch` and ad infinitum.

I used the following script to compare the parsing differences:

```php
<?php

use PhpParser\ParserFactory;
use PhpParser\NodeDumper;

require_once __DIR__ . '/vendor/autoload.php';

$data = '<?php self::$foo["test"]["foo"]();';

$parserFactory = new ParserFactory();
$nodeDumper = new NodeDumper();
// $phpParser = $parserFactory->create(ParserFactory::ONLY_PHP5);
$phpParser = $parserFactory->create(ParserFactory::PREFER_PHP7);
$stmts = $phpParser->parse($data);
echo $nodeDumper->dump($stmts[0]->expr) . PHP_EOL;
```